question/bank/managecategories/category_class.php

```diff
@@ -0,0 +1,609 @@
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class question_category_list extends moodle_list {
+    public $table = "question_categories";
+    public $listitemclassname = 'question_category_list_item';
+    /**
+     * @var reference to list displayed below this one.
+     */
+    public $nextlist = null;
+    /**
+     * @var reference to list displayed above this one.
+     */
+    public $lastlist = null;
+
+    public $context = null;
+    public $sortby = 'parent, sortorder, name';
+
+    public function __construct($type='ul', $attributes='', $editable = false, $pageurl=null, $page = 0, $pageparamname = 'page', $itemsperpage = 20, $context = null){
+        parent::__construct('ul', '', $editable, $pageurl, $page, 'cpage', $itemsperpage);
+        $this->context = $context;
+    }
+
+    public function get_records() {
+        $this->records = get_categories_for_contexts($this->context->id, $this->sortby);
+    }
+
+    /**
+     * Returns the highest category id that the $item can have as its parent.
+     * Note: question categories cannot go higher than the TOP category.
+     *
+     * @param list_item $item The item which its top level parent is going to be returned.
+     * @return int
+     */
+    public function get_top_level_parent_id($item) {
+        // Put the item at the highest level it can go.
+        $topcategory = question_get_top_category($item->item->contextid, true);
+        return $topcategory->id;
+    }
+
+    /**
+     * process any actions.
+     *
+     * @param integer $left id of item to move left
+     * @param integer $right id of item to move right
+     * @param integer $moveup id of item to move up
+     * @param integer $movedown id of item to move down
+     * @return void
+     * @throws coding_exception
+     */
+    public function process_actions($left, $right, $moveup, $movedown) {
+        $category = new stdClass();
+        if (!empty($left)) {
+            // Moved Left (In to another category).
+            $category->id = $left;
+            $category->contextid = $this->context->id;
+            $event = \core\event\question_category_moved::create_from_question_category_instance($category);
+            $event->trigger();
+        } else if (!empty($right)) {
+            // Moved Right (Out of the current category).
+            $category->id = $right;
+            $category->contextid = $this->context->id;
+            $event = \core\event\question_category_moved::create_from_question_category_instance($category);
+            $event->trigger();
+        }
+        parent::process_actions($left, $right, $moveup, $movedown);
+    }
+}
+
+/**
+ * An item in a list of question categories.
+ *
+ * @copyright  1999 onwards Martin Dougiamas {@link http://moodle.com}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class question_category_list_item extends list_item {
```

Comment:
create a different file which will contain only this class and point the code to that, also namespace it against the plugin.